load-testing/ga4gh/drs-performance.js

```diff
@@ -0,0 +1,208 @@
+/*
+Test to simulate a client/user attempting to get x number of signed URLs
+given a list of studies (in the form of `authz` resources).
+
+In other words, we can answer the question "How long does it take to get
+10,000 signed URLs across 3 studies in Gen3?"
+
+The test is not designed with parallelized users
+(there is only 1 virtual user),
+as it's meant to simulate a single person trying to
+get all the data they care about.
+
+**We are assuming the client/user is parallelizing their requests**.
+We accomplish that with k6's `batch` support for requests.
+
+**We also assume the client/user is retrying failed requests**
+(like the rate limiting 503s) with some simple backoff logic
+(for each batch of requests, if any failed, keep retrying the
+failed requests up to 5 times with increasing sleep time in between).
+
+Due to the potential high number of GUIDs that could be requested,
+we are not able to use the existing
+`GUIDS_LIST` that other similar load tests use
+(as it exceeded the max size of commands in linux - 
+it gets passed to a k6 process as a command line arg).
+To circumvent this, the logic of paginating and obtaining the GUIDs
+is built into this test.
+
+There is sufficient flexibility for adjusting num GUIDs,
+num parallel requests, pagination size, authz resources, etc.
+*/
+
+const {
+  check,
+  group,
+  sleep
+} = require('k6'); // eslint-disable-line import/no-unresolved
+const http = require('k6/http'); // eslint-disable-line import/no-unresolved
+const {
+  Rate
+} = require('k6/metrics'); // eslint-disable-line import/no-unresolved
+
+const {
+  TARGET_ENV,
+  AUTHZ_LIST,
+  MINIMUM_RECORDS,
+  RECORD_CHUNK_SIZE,
+  GEN3_HOST,
+  ACCESS_TOKEN,
+  PASSPORTS_LIST,
+  SIGNED_URL_PROTOCOL,
+  NUM_PARALLEL_REQUESTS,
+} = __ENV; // eslint-disable-line no-undef
+
+const authz_list = AUTHZ_LIST.split(',');
+
+const myFailRate = new Rate('failed requests');
+
+export const options = {
+  rps: 90000,
+  thresholds: {
+    http_req_duration: ['avg<3000', 'p(95)<15000'],
+    'failed requests': ['rate<0.1'],
+  },
+  duration: '2h',
+  noConnectionReuse: true,
+  iterations: 1
+};
+
+export default function() {
+  const maxRetries = 5;
+
+  var is_passport_flow = false;
+  var params = {
+    'headers': {
+      'Content-Type': 'application/json',
+      'Authorization': `Bearer ${ACCESS_TOKEN}`,
+    },
+  };
+  var body = null;
+
+  if (PASSPORTS_LIST !== undefined && PASSPORTS_LIST !== null && PASSPORTS_LIST !== "") {
+    console.log(`Passport list supplied: ${PASSPORTS_LIST}. Enabling GA4GH Passport flow to POST list of passports rather than use a Gen3 Access Token.`);
```

Comment:
Please bear in mind that this function will be triggered by potentially dozens / hundreds of VUs (virtual users) which means that it can be executed hundreds of thousands of times , generating exponentially more log entries because of the `console.log(` call... and that results in a few hundred megabytes of Jenkins disk space that will get filled up.

It should be ok if we don't keep too many jobs & console outputs in the history.
Or you could just comment some of them out 🤷🏼‍♂️ 
not a critical thing. It looks good anyway.

Reply:
yeah, good point. At least for this test there are no plans for hundreds of VUs (or to run this regularly). The test is explicitly meant to simulate a single user (I guess we could do multiple in the future, but I don't expect we'd ever get to hundreds). But I definitely get your point